Added automatic clipping of extended attrs when hard clipping.

tfenne opened the pull request.

tfenne commented:
@nh13 I may still tweak this a little and/or add more tests, but this is most of it if you wanted to take a look.

nh13 commented on src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala:
```diff
@@ -48,6 +48,13 @@ import scala.collection.mutable.ListBuffer
+    * be automatically clipped in the same way as the bases and qualities.
+    * @return
+    */
+  def autoClipAttributes: Boolean
```
I am now thinking that `ClippingMode` should also be a `def clippingMode: ClippingMode` just like this, since when are you going to want different clipping modes?  Or am I wrong about that?

nh13 replied:
Looking at a few of the tools, it looks like `mode` is always either hard-clip or soft-clip, so why not the same treatment?

nh13 commented on src/test/scala/com/fulcrumgenomics/bam/SamRecordClipperTest.scala:
```diff
@@ -281,13 +302,34 @@ class SamRecordClipperTest extends UnitSpec {
+      rec.getAttribute("B2") shouldBe Range.inclusive(1, 10).toArray
+    }
+  }
+
```
extra nl

tfenne commented:
@nh13 Good call on the handling of mode vs. autoClip.  I've refactored the trait+object into a more traditional class that just takes both as constructor parameters, and removed the mode parameter from each method.  Changes the tests was a bit of a chore, but I think this is better overall.  Want to take one last quick look?

tfenne commented on src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala:
```diff
@@ -27,12 +27,32 @@ package com.fulcrumgenomics.bam
+  private val NoCallQual = 2.toByte
+}
+
+
```
I know there are two newlines here ... but I like the separation between the object and the class 😛

tfenne commented:
Forgot to re-assign this to you @nh13 when I pushed the review feedback.  It's ready whenever you are for a second pass.

nh13 commented on src/main/scala/com/fulcrumgenomics/bam/ClipOverlappingReads.scala:
```diff
@@ -67,8 +67,7 @@ class ClipOverlappingReads
 
-  private val clippingMode = if (softClip) ClippingMode.Soft else ClippingMode.Hard
-  private val clipper = new SamRecordClipper { override val autoClipAttributes = ClipOverlappingReads.this.autoClipAttributes }
+  private val clipper = new SamRecordClipper(mode=if (softClip) ClippingMode.Soft else ClippingMode.Hard, autoClipAttributes=autoClipAttributes)
```
You could just add an apply method takes a boolean for `softClip`

tfenne replied:
I'd prefer not to do that because a) there are three clipping modes, and b) in the two tools where I use this I have different defaults, and hence different booleans (one is `softClip` the other is `hardClip`).

nh13 commented on src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala:
```diff
@@ -44,17 +64,16 @@ import scala.collection.mutable.ListBuffer
+  *                           and qualities be automatically clipped in the same way as
+  *                           the bases and qualities, otherwise attributes are not
+  *                           touched.
+  * @param mode how should clipping be performed (hard, soft, soft with masking)
```
can you re-order the scaladoc?

nh13 commented on src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala:
```diff
@@ -70,8 +70,7 @@ class TrimPrimers
 )extends FgBioTool with LazyLogging {
-  private val mode = if (hardClip) ClippingMode.Hard else ClippingMode.Soft
-  private val clipper = new SamRecordClipper { override val autoClipAttributes = autoTrimAttributes }
+  private val clipper = new SamRecordClipper(mode=if (hardClip) ClippingMode.Hard else ClippingMode.Soft, autoClipAttributes=autoTrimAttributes)
```
Could use that new `apply` method here too!
